File: 0557-reverse-words-in-a-string-iii/0557-reverse-words-in-a-string-iii.java
class Solution {
    private void reverseCharArray(char[] chars, int start, int end) {
        while (start < end) {
            char tmp = chars[start];
            chars[start] = chars[end];
            chars[end] = tmp;
            start++;
            end--;
        }
    }

    public String reverseWords(String s) {
        char[] chars = s.toCharArray();
        int n = chars.length;
        int start = 0;

        for (int i = 0; i <= n; i++) {
            if (i == n || chars[i] == ' ') {
                reverseCharArray(chars, start, i - 1);
                start = i + 1;
            }
        }

        return new String(chars);
    }
}

File: 0557-reverse-words-in-a-string-iii/0557-reverse-words-in-a-string-iii.kt
class Solution {
    private fun reverseCharArray(chars: CharArray, start: Int, end: Int) {
        var left = start
        var right = end

        while (left < right) {
            val tmp = chars[left]
            chars[left] = chars[right]
            chars[right] = tmp
            left++
            right--
        }
    }

    fun reverseWords(s: String): String {
        val chars = s.toCharArray()
        val n = chars.size
        var start = 0

        for (i in 0..n) {
            if (i == n || chars[i] == ' ') {
                reverseCharArray(chars, start, i - 1)
                start = i + 1
            }
        }

        return String(chars)
    }
}